CIP-0018/CIP-0018.md

```diff
@@ -0,0 +1,104 @@
+- **1.B)** Except for the first stake key (index = 0), stake keys should be registered before being used in output addresses.
+
+The registration of stake keys beyond the first must satisfy one of the following conditions:
+
+- **2.A)** It must exist an active<sup><a href="#1-active-stake-key">1</a></sup> stake key registration certificate on chain of the immediately previous stake key, located in the stable area<sup><a href="#2-stable-area">2</a></sup> of the chain. 
+- **2.B)** It must exist another stake key registration certificate of the immediately previous stake key in the same transaction which satisfies 2.A or 2.B.
+
+Said differently, one can only register a new stake key if previous keys have already been registered and are immutable. We also allow registering keys in batch to alleviate the user experience for users wanting to register many stake keys.
+
+> **For example**, If a slot 10, a wallet registers a stake key at index = 0. At slot 255500, the wallet registers keys from index = 1 to index = 10 in the same transaction. This is fine because the key registration at index 0 can be considered _stable_ and _immutable_ (it's part of all possible chain forks) and the next 10 keys are registered as part of a single transaction. 
+
+Note that the condition 1.B) is only a best-effort to prevent so-called mangled addresses<sup><a href="#3-mangled-addresses">3</a></sup>. It is possible for a key registration to be rolled back, and for an address used later to point to a non-existing key (from the perspective of the main chain). We acknowledge this limitation under the assumption that users or wallets would eventually re-submit any rolled back registration (which are retro-active). 
+
+However, the condition 2.A) is a hard limitation for the good functioning of this specification for it enforces the invariant stated in introduction: 
+
+> **(invariant)** At all time, the key indexes must form a contiguous sequence with no gap. 
+
+This invariant makes it possible for wallets to know when they should generate and look for new stake keys indexes. It allows for efficient implementations both in terms of space and computational cost. Without waiting for stabilisation of certificates, some intermediate certificate may get lost in translation because of chain forks / rollbacks. By ensuring that any previous certificate is active and stable, we ensure that the associated part of the HD tree (that is for an account `i` `1852'/1815'/i'/2/*`) can be recovered fully from the chain at a later time; it requires however a linear scan of the chain.
+
+### Key de-registration
+
+Similarly, keys can only be de-registered in sequence, from the latest and downwards. Transaction de-registering stake keys must satisfy one of the following conditions to be valid:
+
+- **3.A)** The index of the de-registered key must be equal to the index of the last active<sup><a href="#1-active-stake-key">1</a></sup> stake key on chain.
```

Comment:
Should also note this can cause backwards compatibility issues since Yoroi supports de-registering your stake key (so if it doesn't support this spec, it will de-register only the 0-index key, leaving the other active keys untouched breaking this rule)

Reply:
Run true 
end point code separated by 18 Dec hex point walletConnect